src/preamble.js

```diff
@@ -2179,6 +2179,9 @@ function integrateWasmJS() {
       if (exports.memory) mergeMemory(exports.memory);
       Module['asm'] = exports;
       Module["usingWasm"] = true;
+#if WASM_BACKEND
+      Module["_stackRestore"](STACKTOP);
```

Comment:
Hmm, this doesn't look right to me. First, I think it should be exported without the `_` prefix. Also, it isn't necessarily exported (we no longer export all the runtime by default). Instead, calling `stackRestore` directly (not on Module) should work, like it was before - why did you change it from that? I may be missing something here.

Reply:
Hmm.. I changed it because I was getting a few test failure that looked like this:

```
+TypeError: stackRestore is not a function
+    at receiveInstance (/tmp/tmpyg29kY/emscripten_test_binaryen2_aNCc2n/src.cpp.o.js:1:17965)
+    at doNativeWasm (/tmp/tmpyg29kY/emscripten_test_binaryen2_aNCc2n/src.cpp.o.js:1:18777)
+    at Object.Module.asm (/tmp/tmpyg29kY/emscripten_test_binaryen2_aNCc2n/src.cpp.o.js:1:20161)
+    at Object.<anonymous> (/tmp/tmpyg29kY/emscripten_test_binaryen2_aNCc2n/src.cpp.o.js:1:106527)
+    at Module._compile (module.js:573:30)
+    at Object.Module._extensions..js (module.js:584:10)
+    at Module.load (module.js:507:32)
+    at tryModuleLoad (module.js:470:12)
+    at Function.Module._load (module.js:462:3)
+    at Function.Module.runMain (module.js:609:10)
```

e.g. test_relocatable_void_function

Reply:
Turns out the other approach also gives test failures too. so I don't have a working solution yet.

Reply:
Hmm, in emscripten.py I see
```
  # some runtime functionality may not have been generated in
  # the wasm; provide a JS shim for it
  for name in ['setTempRet0', 'getTempRet0', 'stackSave', 'stackRestore', 'stackAlloc']:
    if name not in exported_implemented_functions:
      module.append('var %s;\n' % name)
```
The last part generates a `var stackRestore` is the wasm doesn't contain it (necessary for closure compiler). Which suggests that the wasm backend path doesn't always generate it - not sure why that would be?

Reply:
With -O0:

```
/tmp/emscripten_test_binaryen0_NqSohl/src.cpp.o.js:110
+      throw ex;
+      ^
+abort("cannot use the stack before compiled code is ready to run, and has provided stack access") at Error
+    at jsStackTrace (/tmp/emscripten_test_binaryen0_NqSohl/src.cpp.o.js:1190:13)
+    at stackTrace (/tmp/emscripten_test_binaryen0_NqSohl/src.cpp.o.js:1207:12)
+    at abort (/tmp/emscripten_test_binaryen0_NqSohl/src.cpp.o.js:2811:44)
+    at getTempRet0 (/tmp/emscripten_test_binaryen0_NqSohl/src.cpp.o.js:226:3)
+    at receiveInstance (/tmp/emscripten_test_binaryen0_NqSohl/src.cpp.o.js:1821:7)
+    at doNativeWasm (/tmp/emscripten_test_binaryen0_NqSohl/src.cpp.o.js:1848:5)
+    at Object.Module.asm (/tmp/emscripten_test_binaryen0_NqSohl/src.cpp.o.js:1927:15)
+    at Object.<anonymous> (/tmp/emscripten_test_binaryen0_NqSohl/src.cpp.o.js:2237:24)
+    at Module._compile (module.js:573:30)
+    at Object.Module._extensions..js (module.js:584:10)

```

Reply:
Seems like it fails when BINARYEN_ASYNC_COMPILATION is off and some tests force this to be off..

Reply:
Ah good, that `cannot use the stack before compiled code is ready to run` assertion in O0 may explain what's going on here. We are still in the process of setting things up for compiled code, and calling into it here may be dangerous. In practice it isn't here, so we can either add an exception for this function, or we can do this call in a later time when the runtime is ready. I think that's best, with something like this:
```
addOnInit(function() {
  stackRestore(STACKTOP);
});
```